no-literal-checker/src/main/java/org/checkerframework/checker/noliteral/NoLiteralAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,188 @@
+   * anything...
+   */
+  private class NoLiteralQualifierDefaults extends QualifierDefaults {
+    public NoLiteralQualifierDefaults(
+        Elements elements, NoLiteralAnnotatedTypeFactory noLiteralAnnotatedTypeFactory) {
+      super(elements, noLiteralAnnotatedTypeFactory);
+    }
+
+    /** Add standard unchecked defaults that do not conflict with previously added defaults. */
+    @Override
+    public void addUncheckedStandardDefaults() {
+      // do nothing
+    }
+
+    /**
+     * Replace the normal default applier element with one that also applies the defaults to the
+     * component types of arrays. For the no-literal checker, it makes a lot more sense to talk
+     * about "a non-constant array of string constants" than it does to say "a constant array of
+     * non-constant strings" -- the latter doesn't even make sense. So to do optimistic defaulting
+     * in bytecode correctly, the defaults need to be applied all the way down through an array.
+     */
+    @Override
+    protected DefaultApplierElement createDefaultApplierElement(
+        AnnotatedTypeFactory atypeFactory,
+        Element annotationScope,
+        AnnotatedTypeMirror type,
+        boolean applyToTypeVar) {
+      // Must check because the defaults for source code must not change to avoid interfering with
+      // CLIMB-to-top local type inference.
+      if (ElementUtils.isElementFromByteCode(annotationScope)) {
+        return new NoLiteralDefaultApplierElement(
+            atypeFactory, annotationScope, type, applyToTypeVar);
+      } else {
+        return super.createDefaultApplierElement(
+            atypeFactory, annotationScope, type, applyToTypeVar);
+      }
+    }
+
+    /**
+     * Modification of the DefaultApplierElement to also apply defaults to array component types.
+     *
+     * <p>See summary comment above for reasoning.
+     */
+    private class NoLiteralDefaultApplierElement extends DefaultApplierElement {
+
+      public NoLiteralDefaultApplierElement(
+          AnnotatedTypeFactory atypeFactory,
+          Element annotationScope,
+          AnnotatedTypeMirror type,
+          boolean applyToTypeVar) {
+        super(atypeFactory, annotationScope, type, applyToTypeVar);
+      }
+
+      /**
+       * Add the qualifier to the type if it does not already have an annotation in the same
+       * hierarchy as qual.
+       *
+       * @param type type to add qual
+       * @param qual annotation to add
+       */
+      @Override
+      protected void addAnnotation(AnnotatedTypeMirror type, AnnotationMirror qual) {
+        super.addAnnotation(type, qual);
+        if (type.getKind() == TypeKind.ARRAY) {
+          AnnotatedArrayType asArrayType = (AnnotatedArrayType) type;
+          addAnnotation(asArrayType.getComponentType(), qual);
+        }
+      }
+    }
+  }
+
+  @Override
+  protected void checkInvalidOptionsInferSignatures() {
+    // This checker is specifically designed to work with whole-program inference,
```

Comment:
This comment is confusing, since the whole point of whole-program inference is that it works with every type system, without change.  Can you rephrase it to be more specific?

Reply:
I'm not sure how I could make this more specific. Whole program inference refuses to run with certain command line options that this checker requires (in particular, `-AuseConservativeDefaults=-bytecode`, see `NoLiteralChecker.java`). This comment says that this checker has been specifically tested with WPI and works, so the check is turned off. I don't know why the check was there in the first place, to be honest, so it could possibly be removed.